Core/GCUniqueCommitCollection.m

```diff
@@ -0,0 +1,71 @@
+//  Copyright (C) 2015 Pierre-Olivier Latour <info@pol-online.net>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#import "GCUniqueCommitCollection.h"
+
+#import "GCCommit.h"
+
+
+@implementation GCUniqueCommitCollection {
+  NSMutableArray* _commits; // Contains all the commits, even removed ones
+  NSMutableSet* _actualCommitHashes; // Commits that were added but have not been removed
+  NSMutableSet* _removedCommitHashes;
+}
+
+- (instancetype)init {
+  if ((self = [super init])) {
+    _commits = [NSMutableArray array];
+    _actualCommitHashes = [NSMutableSet new];
+    _removedCommitHashes = [NSMutableSet new];
+  }
+  return self;
+}
+
+- (void)addCommit:(GCCommit *)commit {
+  if (![self hasCommit:commit]) {
+    if ([_removedCommitHashes containsObject:commit.SHA1]) {
+      [_removedCommitHashes removeObject:commit.SHA1];
+    } else {
+      [_commits addObject:commit];
+    }
+    [_actualCommitHashes addObject:commit.SHA1];
+  }
+}
+
+- (void)removeCommit:(GCCommit*)commit {
+  if (![self hasCommit:commit]) {
```

Comment:
Use consistent patterns: either `if (!...) return` or `if (...) { }` but don't mix and match as you are doing with `-addCommit:` and `-removeCommit:`.